### ISSUE TITLE
How to debug with python operator?

### ISSUE THREAD

**Issue by zdltheone** [dmlc/mxnet#2447]
I'm trying to write some operators with python and at first, I tried "softmax" with NDarray, but there's some error occured. The following is my code :
````
class NDsoftmax(mx.operator.NDArrayOp):
    def __init__(self):
        super(NDsoftmax, self).__init__(False)
        self.fwd_kernel = None
        self.bwd_kernel = None
        
```
    def list_arguments(self):
        return ['data', 'label']

    def list_outputs(self):
        return ['output']

    def infer_shape(self, in_shape):
        data_shape = in_shape[0]
        label_shape = (in_shape[0][0],)
        output_shape = in_shape[0]
        return [data_shape, label_shape], [output_shape]
    
    def forward(self, in_data, out_data):
        '''
       #this works but only use cpu
        x = in_data[0].asnumpy()
        y = np.exp(x - x.max(axis=1).reshape((x.shape[0], 1)))
        y /= y.sum(axis=1).reshape((x.shape[0], 1))
        z = out_data[0] 
        mx.nd.array(y).copyto(z)
        '''
        #doesn't work and don't know why
        x = in_data[0]
        y = out_data[0]
        y[:] = mx.nd.exp(x - mx.nd.max(x, axis = 1).reshape((x.shape[0],1)))
        y = y / (mx.nd.sum(y, axis = 1).reshape((x.shape[0],1)))
        
        #out_data[0] = mx.nd.array(y).copy()
        #mx.nd.array(y).copyto(z)
        
        #y = out_data[0]
    def backward(self, out_grad, in_data, out_data, in_grad):
        l = in_data[1].asnumpy().ravel().astype(np.int)
        y = out_data[0].asnumpy()
        y[np.arange(l.shape[0]), l] -= 1.0
        z = in_grad[0] 
        mx.nd.array(y).copyto(z)
        #self.assign(in_grad[0], req[0], mx.nd.array(y))

```
data = mx.symbol.Variable('data')
fc1  = mx.symbol.FullyConnected(data = data, num_hidden=128)
act1 = mx.symbol.Activation(data = fc1, act_type="relu")
fc2  = mx.symbol.FullyConnected(data = act1, num_hidden = 64)
act2 = mx.symbol.Activation(data = fc2, act_type="relu")
fc3  = mx.symbol.FullyConnected(data = act2, num_hidden=10)
mysoftmax = NDsoftmax()
mlp  = mysoftmax(data=fc3, name='softmax')
#mlp  = mx.symbol.Custom(data=fc3, name='softmax', op_type='Softmax_cpu')

logging.basicConfig(level=logging.DEBUG)


train = mx.io.MNISTIter(
    image      = "example/image-classification/mnist/train-images-idx3-ubyte",
    label      = "example/image-classification/mnist/train-labels-idx1-ubyte",
    batch_size = 128,
    data_shape = (784, ))
val   = mx.io.MNISTIter(
    image      = "example/image-classification/mnist/t10k-images-idx3-ubyte",
    label      = "example/image-classification/mnist/t10k-labels-idx1-ubyte",
    batch_size = 128,
    data_shape = (784, ))


model = mx.model.FeedForward(
    ctx = mx.gpu(0), symbol = mlp, num_epoch = 20,
    learning_rate = 0.1, momentum = 0.9, wd = 0.00001)

model.fit(X=train, eval_data=val,
          batch_end_callback=mx.callback.Speedometer(100,100))
````

It seems something wrong with the forward function as the accuracy never increase, and I don't  know how to debug because when I add a breakpoint at " forward " function, it won't step into the function.

**Comment by Godricly** [comment]
check [here](https://github.com/dmlc/mxnet/blob/master/tests/python/unittest/test_operator.py)  to one layer alone. Also the NDArrayOp is deprecated. Try newer [CustomOp](http://mxnet.readthedocs.io/en/latest/how_to/new_op.html) if you can. example [link](https://github.com/dmlc/mxnet/tree/master/example/numpy-ops).

**Comment by zdltheone** [comment]
ok, I'll try CustomOp, thank you :)  @Godricly
